Archive/Imo/Imo2019Q4.lean

```diff
@@ -41,41 +42,48 @@ namespace Imo2019Q4
 
 theorem upper_bound {k n : ℕ} (hk : k > 0)
     (h : (k ! : ℤ) = ∏ i in range n, ((2:ℤ) ^ n - (2:ℤ) ^ i)) : n < 6 := by
-  have h2 : n * (n - 1) / 2 < k := by
-    suffices multiplicity 2 (k ! : ℤ) = (n * (n - 1) / 2 : ℕ) by
+  have h2 : (∑ i in range n, i) < k
+  · suffices multiplicity 2 (k ! : ℤ) = ↑(∑ i in range n, i : ℕ) by
       rw [← PartENat.coe_lt_coe, ← this]; change multiplicity ((2 : ℕ) : ℤ) _ < _
       simp_rw [Int.coe_nat_multiplicity, multiplicity_two_factorial_lt hk.lt.ne.symm]
-    rw [h, multiplicity.Finset.prod Int.prime_two, ← sum_range_id, Nat.cast_sum]
+    rw [h, multiplicity.Finset.prod Int.prime_two, Nat.cast_sum]
     apply sum_congr rfl; intro i hi
     rw [multiplicity_sub_of_gt, multiplicity_pow_self_of_prime Int.prime_two]
     rwa [multiplicity_pow_self_of_prime Int.prime_two, multiplicity_pow_self_of_prime Int.prime_two,
       PartENat.coe_lt_coe, ← mem_range]
   rw [← not_le]; intro hn
-  apply _root_.ne_of_lt _ h.symm
-  suffices (∏ i in range n, ↑2 ^ n : ℤ) < ↑k ! by
-    apply lt_of_le_of_lt _ this; apply prod_le_prod
-    · intros; rw [sub_nonneg]; apply pow_le_pow; norm_num; apply le_of_lt; rwa [← mem_range]
-    · intros; apply sub_le_self; apply pow_nonneg; norm_num
-  suffices 2 ^ (n * n) < (n * (n - 1) / 2)! by
-    rw [prod_const, card_range, ← pow_mul]; rw [← Int.ofNat_lt] at this
-    clear h; convert this.trans _ using 1; norm_cast; rwa [Int.ofNat_lt, factorial_lt]
-    refine' Nat.div_pos _ (by norm_num)
-    refine' le_trans _ (mul_le_mul hn (pred_le_pred hn) (zero_le _) (zero_le _))
+  apply _root_.ne_of_gt _ h
+  have le_sum : ∀ N ≥ 6, 15 ≤ ∑ i in range N, i
+  · intro N hN
+    show ∑ i in range 6, i ≤ _
+    apply sum_le_sum_of_subset
+    simpa using hN
+  calc ∏ i in range n, ((2:ℤ) ^ n - (2:ℤ) ^ i) ≤ ∏ i in range n, (2:ℤ) ^ n := ?_
+    _ < ↑ k ! := ?_
+  · gcongr
+    · intro i hi
+      replace hi : i < n := by simpa using hi
+      have : (2:ℤ) ^ i ≤ (2:ℤ) ^ n := by gcongr; norm_num
+      linarith
+    · apply sub_le_self
+      positivity
+  norm_cast
+  calc ∏ i in range n, 2 ^ n = 2 ^ (n * n) := by rw [prod_const, card_range, ← pow_mul]
+    _ < (∑ i in range n, i)! := ?_
+    _ < k ! := by have := le_sum n hn; gcongr
+  clear h h2
+  induction' n, hn using Nat.le_induction with n' hn' IH
+  · show _ < 15!
```

Comment:
this needs a porting note, I think it can be removed after we have ported [`algebra.big_operators.norm_num`](https://leanprover-community.github.io/mathlib-port-status/file/algebra/big_operators/norm_num)

Reply:
Ah, in fact `norm_num` does it already. Thanks!